File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/others/distance/BestNearest.java
package cujae.inf.ic.om.heuristic.assignment.others.distance;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.heuristic.assignment.classical.ByNotUrgency;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

/*Clase que modela como asignar los clientes a los depósitos partiendo del criterio de ser los mejores candidatos*/
public class BestNearest extends ByNotUrgency {

	private Solution solution = new Solution();
	
	private ArrayList<Cluster> listClusters;
	private ArrayList<Customer> listCustomersToAssign;
	private ArrayList<Integer> listIDDepots;
	private NumericMatrix costMatrix;
	
    public BestNearest() {
        super();
    }
    
    @Override
    public Solution toClustering() {
    	initialize();
    	assign();
    	return finish();
    }
	
    @Override
    public void initialize() {
    	listClusters = initializeClusters();	
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		listIDDepots = new ArrayList<Integer>(Problem.getProblem().getListIDDepots());  
		costMatrix = initializeCostMatrix(listCustomersToAssign, Problem.getProblem().getDepots(), distanceType);
    }
		
	@Override
	public void assign() {
		int idCustomer = -1;
		int posCustomer = -1;
		double requestCustomer = 0.0;
		
		int idDepot = -1;
		int posDepot = -1;
		double capacityDepot = 0.0;
		
		int posCluster = -1; 
		double requestCluster = 0.0;
		
		RowCol rcBestAll = null;
		
		int totalCustomers = listCustomersToAssign.size();
		int totalDepots = listIDDepots.size(); 
		
        while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty()) && (!costMatrix.fullMatrix(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1), Double.POSITIVE_INFINITY))) 
        {					
        	rcBestAll = costMatrix.indexLowerValue(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1));

        	posDepot = rcBestAll.getRow() - totalCustomers;
        	idDepot = listIDDepots.get(posDepot);      	
        	capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idDepot));

           	posCustomer = rcBestAll.getCol();
           	idCustomer = Problem.getProblem().getCustomers().get(posCustomer).getIDCustomer();
        	requestCustomer = Problem.getProblem().getCustomers().get(posCustomer).getRequestCustomer();

        	posCluster = findCluster(idDepot, listClusters);
        	
        	if(posCluster != -1)
        	{
        		requestCluster = listClusters.get(posCluster).getRequestCluster();
            	
        		if(capacityDepot >= (requestCluster + requestCustomer)) 
        		{
        			requestCluster += requestCustomer;
        			
        			listClusters.get(posCluster).setRequestCluster(requestCluster);
        			listClusters.get(posCluster).getItemsOfCluster().add(idCustomer);

        			costMatrix.fillValue(totalCustomers, posCustomer, (totalCustomers + totalDepots - 1), posCustomer, Double.POSITIVE_INFINITY);
        			listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomer));
        		} 
        		else
        			costMatrix.setItem(rcBestAll.getRow(), posCustomer, Double.POSITIVE_INFINITY);
        		
        		if(isFullDepot(listCustomersToAssign, requestCluster, capacityDepot))
    			{
					if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
						solution.getClusters().add(listClusters.remove(posCluster));
					else
						listClusters.remove(posCluster);
    			
    				costMatrix.fillValue(rcBestAll.getRow(), 0, rcBestAll.getRow(), (totalCustomers + totalDepots - 1), Double.POSITIVE_INFINITY);
    			}
        	}
        }
	}
       
    @Override
    public Solution finish() {
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));
		
		OSRMService.clearDistanceCache();
    	
		return solution;
    }
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/others/distance/best_nearest.py
import numpy as np
from typing import List
from ...classical.by_not_urgency import ByNotUrgency
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.output.solution import Solution
from ....problem.output.cluster import Cluster

class BestNearest(ByNotUrgency):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        self.list_id_depots: List[int]
        
    def to_clustering(self):
        self.initialize()
        self.assign()
        return self.finish()
    
    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_id_depots = list(Problem.get_problem().get_list_id_depots())
        
    def assign(self):
        cost_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        
        while any(self.list_customers_to_assign) and self.list_clusters:
            min_value = np.min(cost_matrix) 
            row_best, col_best = np.where(cost_matrix == min_value)
            row_best, col_best = row_best[0], col_best[0]

            selected_customer = Problem.get_problem().get_customers()[col_best]
            request_customer = selected_customer.get_request_customer()
              
            id_depot = Problem.get_problem().get_list_id_depots()[row_best]    
            selected_depot: Depot = Problem.get_problem().get_depot_by_id_depot(id_depot)
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(selected_depot)
                
            pos_cluster = self.find_cluster(id_depot, self.list_clusters)

            if pos_cluster != -1:
                cluster: Cluster = self.list_clusters[pos_cluster]
                request_cluster = cluster.get_request_cluster()
        
                if capacity_depot >= (request_cluster + request_customer):
                    cluster.set_request_cluster(request_cluster + request_customer)
                    cluster.get_items_of_cluster().append(selected_customer.get_id_customer())

                    capacity_depot -= request_customer
                    self.list_customers_to_assign[col_best] = None
                    
                    if self.is_full_depot(self.list_customers_to_assign, cluster.get_request_cluster(), capacity_depot):
                        if cluster.get_items_of_cluster():
                            self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                    
                    cost_matrix[:, col_best] = float('inf')    
            
        if np.all(cost_matrix == float("inf")):
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_total_unassigned_items().append(selected_customer.get_id_customer())

              
    def finish(self):
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_unassigned_items().append(customer.get_id_customer())
            
        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)
                        
        OSRMService.clear_distance_cache()
        
        return self.solution